### PR TITLE
fix(router): do not finish bootstrap until all the routes are resolved

### PR DESCRIPTION
**Pull request by vsavkin**
**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

The bootstrap doesn't wait for the router to finish the initial navigation.

**What is the new behavior?**

There is an option you can pass to RouterModule.foRoot to make bootstrap wait. Since many depend on the current behavior, it is preserved as a default and is deprecated.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: modules/@angular/router/src/router_module.ts [hunk 1]
```diff
@@ -208,6 +210,26 @@ export function provideRoutes(routes: Routes): any {
...
+ *
+ * @description
+ * * 'enabled' - the initial navigation starts before the root component is created.
+ * The bootstrap is blocked until the initial navigation is complete.
+ * * 'disabled' - the initial navigation is not performed. The localtion listener is set up before
+ * the root component gets created.
+ * * 'legacy_enabled'- the initial navigation starts after the root component has been created.
+ * The bootstrap is not blocked until the initial navigation is complete. @deprecated
+ * * 'legacy_disabled'- the initial navigation is not performed. The localtion listener is set up
+ * after @deprecated
```
Comment (vicb) [inline modules/@angular/router/src/router_module.ts]: - typos in "localtion"
- could you be more verbose on what the preferred choice should be (ie most probably `"enabled"`) and when you would like to switch to "disable". Also say that the "legacy_..." options are for backward compatibility and should not be used for new apps ?

File: modules/@angular/router/src/router_module.ts [hunk 2]
```diff
@@ -208,6 +210,26 @@ export function provideRoutes(routes: Routes): any {
...
+ * the root component gets created.
+ * * true - same as 'legacy_enabled'. @deprecated
```
Comment (vicb) [inline modules/@angular/router/src/router_module.ts]: back quote true/false

File: modules/@angular/router/src/router_module.ts [hunk 3]
```diff
@@ -278,22 +300,100 @@ export function rootRoute(router: Router): ActivatedRoute {
...
     }
-  };
+
+    preloader.setUpPreloading();
+    router.resetRootComponentType(ref.componentTypes[0]);
+    this.resultOfPreactivationDone.next(null);
+    this.resultOfPreactivationDone.complete();
+  }
+
+  isLegacyEnabled(opts: ExtraOptions): boolean {
```
Comment (vicb) [inline modules/@angular/router/src/router_module.ts]: move these 2 mehtods out of the class ?

File: modules/@angular/router/src/router_module.ts [hunk 4]
```diff
@@ -278,22 +300,100 @@ export function rootRoute(router: Router): ActivatedRoute {
...
+ * a resolver or a guards executes asynchronously. Second, we need to actually run
+ * activation in a BOOTSTRAP_LISTENER. We utilize the afterPreactivation
+ * hook provided by the router to do that.
+ *
+ * The router navigation starts, reaches the point when preactivation is done, and then
+ * pauses. It waits for the hook to be resolved. We then resolve it only in a bootstrap listener.
+ */
+@Injectable()
+export class RouterInitializer {
+  private initNavigation: boolean;
```
Comment (vicb) [inline modules/@angular/router/src/router_module.ts]: `initNavigationDone = false` ?

**Comment by vicb** [comment]
Please add a `DEPRECATIONS` section in the commit message. Thanks !

**Review by vicb — APPROVED** [review]

File: tools/public_api_guard/router/typings/router.d.ts [hunk 5]
```diff
@@ -76,7 +76,7 @@ export declare type Event = NavigationStart | NavigationEnd | NavigationCancel |
 export interface ExtraOptions {
     enableTracing?: boolean;
     errorHandler?: ErrorHandler;
-    initialNavigation?: boolean;
+    initialNavigation?: InitialNavigation;
```
Comment (IgorMinar) [inline tools/public_api_guard/router/typings/router.d.ts]: isn't this a breaking change? when was this interface introduced?

Comment (vicb) [inline tools/public_api_guard/router/typings/router.d.ts]: see https://github.com/angular/angular/pull/14762/files#diff-c0baae5e1df628e1a217e8dc38557fcbR237

the changes are described in the commit comment: former options are valid and have the same meaning (`true`/`false`), new values have a new meaning and are documented,

**Review by IgorMinar — APPROVED** [review]

**Pull request merged**